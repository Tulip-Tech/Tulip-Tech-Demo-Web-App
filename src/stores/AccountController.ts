import jwtDecode from 'jwt-decode';

import { useFavoritesStore } from '#src/stores/FavoritesStore';
import { useWatchHistoryStore } from '#src/stores/WatchHistoryStore';
import type {
  AuthData,
  Capture,
  Customer,
  CustomerConsent,
  GetCaptureStatusResponse,
  GetCustomerConsentsResponse,
  GetPublisherConsentsResponse,
  JwtDetails,
} from '#types/account';
import { useConfigStore } from '#src/stores/ConfigStore';
import * as persist from '#src/utils/persist';
import { useAccountStore } from '#src/stores/AccountStore';
import { restoreWatchHistory, serializeWatchHistory } from '#src/stores/WatchHistoryController';
import { restoreFavorites, serializeFavorites } from '#src/stores/FavoritesController';
import { getMediaByWatchlist } from '#src/services/api.service';
import { queryClient } from '#src/containers/QueryProvider/QueryProvider';
import useService from '#src/hooks/useService';
import useAccount from '#src/hooks/useAccount';

const PERSIST_KEY_ACCOUNT = 'auth';

let subscription: undefined | (() => void);
let refreshTimeout: number;

export const authNeedsRefresh = (auth: AuthData): boolean => {
  const decodedToken: JwtDetails = jwtDecode(auth.jwt);
  const expiresIn = decodedToken.exp * 1000 - Date.now();

  // returns true if token expires in 5 minutes or less
  return expiresIn < 5 * 60 * 1000;
};

export const setJwtRefreshTimeout = () => {
  const auth = useAccountStore.getState().auth;

  // if jwp integration, skip code below
  if (!auth?.refreshToken) return;

  window.clearTimeout(refreshTimeout);

  if (auth && !document.hidden) {
    refreshTimeout = window.setTimeout(() => refreshJwtToken(auth), 60 * 5 * 1000);
  }
};

export const handleVisibilityChange = () => {
  if (document.hidden) return window.clearTimeout(refreshTimeout);

  // document is visible again, test if we need to renew the token
  const auth = useAccountStore.getState().auth;

  // user is not logged in / if jwp integration, skip code below
  if (!auth || !auth?.refreshToken) return;

  // refresh the jwt token if needed. This starts the timeout as well after receiving the refreshed tokens.
  if (authNeedsRefresh(auth)) return refreshJwtToken(auth);

  // make sure to start the timeout again since we've cleared it when the document was hidden.
  setJwtRefreshTimeout();
};

export const initializeAccount = async () => {
  await useService(async ({ accountService, config }) => {
    useAccountStore.setState({
      loading: true,
      canUpdateEmail: accountService.canUpdateEmail,
      canRenewSubscription: accountService.canRenewSubscription,
      canChangePasswordWithOldPassword: accountService.canChangePasswordWithOldPassword,
    });
    accountService.setEnvironment(config);

    const storedSession: AuthData | null = persist.getItem(PERSIST_KEY_ACCOUNT) as AuthData | null;

    // clear previous subscribe (for dev environment only)
    if (subscription) {
      subscription();
    }

    document.removeEventListener('visibilitychange', handleVisibilityChange);
    document.addEventListener('visibilitychange', handleVisibilityChange);

    subscription = useAccountStore.subscribe(
      (state) => state.auth,
      (authData) => {
        setJwtRefreshTimeout();
        persist.setItem(PERSIST_KEY_ACCOUNT, authData);
      },
    );

    // restore session from localStorage
    try {
      if (storedSession) {
        const refreshedAuthData = await accountService.getFreshJwtToken({ config, auth: storedSession });
        if (refreshedAuthData) {
          await getAccount(refreshedAuthData);
          await restoreWatchHistory();
          await restoreFavorites();
        }
      }
    } catch (error: unknown) {
      await logout();
    }

    useAccountStore.setState({ loading: false });
  });
};

export async function updateUser(
  values: { firstName: string; lastName: string } | { email: string; confirmationPassword: string },
): Promise<ServiceResponse<Customer>> {
  return await useService(async ({ accountService, sandbox = true }) => {
    useAccountStore.setState({ loading: true });

    const { auth, user, canUpdateEmail } = useAccountStore.getState();

    if (Object.prototype.hasOwnProperty.call(values, 'email') && !canUpdateEmail) {
      throw new Error('Email update not supported');
    }

    if (!auth || !user) {
      throw new Error('no auth');
    }

    const response = await accountService.updateCustomer({ ...values, id: user.id.toString() }, sandbox, auth.jwt);

    if (!response) {
      throw new Error('Unknown error');
    }

    if (response.errors?.length === 0) {
      useAccountStore.setState({ user: response.responseData });
    }

    return response;
  });
}

export const refreshJwtToken = async (auth: AuthData) => {
  await useService(async ({ accountService }) => {
    try {
      const { config } = useConfigStore.getState();
      const authData = await accountService.getFreshJwtToken({ config, auth });

      if (authData) {
        useAccountStore.setState((s) => ({ auth: { ...s.auth, ...authData } }));
      }
    } catch (error: unknown) {
      // failed to refresh, logout user
      await logout();
    }
  });
};

export const getAccount = async (auth: AuthData) => {
  await useService(async ({ accountService, config, accessModel }) => {
    const response = await accountService.getUser({ config, auth });
    if (response) {
      await afterLogin(auth, response.user, response.customerConsents, accessModel);
    }

    useAccountStore.setState({ loading: false });
  });
};

export const login = async (email: string, password: string) => {
  await useService(async ({ accountService, config, accessModel }) => {
    useAccountStore.setState({ loading: true });

    const response = await accountService.login({ config, email, password });
    if (response) {
      await afterLogin(response.auth, response.user, response.customerConsents, accessModel);

      await restoreFavorites();
      await restoreWatchHistory();
    }

    useAccountStore.setState({ loading: false });
  });
};

export async function logout() {
  if (!useAccountStore.getState().auth) return;

  await useService(async ({ accountService }) => {
    persist.removeItem(PERSIST_KEY_ACCOUNT);

    // this invalidates all entitlements caches which makes the useEntitlement hook to verify the entitlements.
    await queryClient.invalidateQueries('entitlements');

    useAccountStore.setState({
      auth: null,
      user: null,
      subscription: null,
      transactions: null,
      activePayment: null,
      customerConsents: null,
      publisherConsents: null,
    });

    await restoreFavorites();
    await restoreWatchHistory();

    // it's needed for the InPlayer SDK
    await accountService?.logout();
  });
}

export const register = async (email: string, password: string) => {
  await useService(async ({ accountService, accessModel, config }) => {
    useAccountStore.setState({ loading: true });
    const response = await accountService.register({ config, email, password });
    if (response) {
      const { auth, user, customerConsents } = response;
      await afterLogin(auth, user, customerConsents, accessModel, false);
    }

    await updatePersonalShelves();
  });
};

export const updatePersonalShelves = async () => {
  await useAccount(async ({ customer, auth: { jwt } }) => {
    await useService(async ({ accountService, sandbox = true }) => {
      const { watchHistory } = useWatchHistoryStore.getState();
      const { favorites } = useFavoritesStore.getState();
      if (!watchHistory && !favorites) return;

      const personalShelfData = {
        history: serializeWatchHistory(watchHistory),
        favorites: serializeFavorites(favorites),
      };

      return await accountService?.updatePersonalShelves(
        {
          id: customer.id,
          externalData: personalShelfData,
        },
        sandbox,
        jwt,
      );
    });
  });
};

export const updateConsents = async (customerConsents: CustomerConsent[]): Promise<ServiceResponse<CustomerConsent[]>> => {
  return await useAccount(async ({ customer, auth: { jwt } }) => {
    return await useService(async ({ accountService, config }) => {
      useAccountStore.setState({ loading: true });

      try {
        const response = await accountService?.updateCustomerConsents({
          jwt,
          config,
          customer,
          consents: customerConsents,
        });

        if (response?.consents) {
          useAccountStore.setState({ customerConsents: response.consents });
          return {
            responseData: response.consents,
            errors: [],
          };
        }
        return {
          responseData: [],
          errors: [],
        };
      } finally {
        useAccountStore.setState({ loading: false });
      }
    });
  });
};

// TODO: Decide if it's worth keeping this or just leave combined with getUser
// noinspection JSUnusedGlobalSymbols
export async function getCustomerConsents(): Promise<GetCustomerConsentsResponse> {
  return await useAccount(async ({ customer, auth: { jwt } }) => {
    return await useService(async ({ accountService, config }) => {
      const response = await accountService.getCustomerConsents({ config, customer, jwt });

      if (response?.consents) {
        useAccountStore.setState({ customerConsents: response.consents });
      }

      return response;
    });
  });
}

export const getPublisherConsents = async (): Promise<GetPublisherConsentsResponse> => {
  return await useService(async ({ accountService, config }) => {
    const response = await accountService.getPublisherConsents(config);

    useAccountStore.setState({ publisherConsents: response.consents });

    return response;
  });
};

export const getCaptureStatus = async (): Promise<GetCaptureStatusResponse> => {
  return await useAccount(async ({ customer, auth: { jwt } }) => {
    return await useService(async ({ accountService, sandbox = true }) => {
      const { responseData } = await accountService.getCaptureStatus({ customer }, sandbox, jwt);

      return responseData;
    });
  });
};

export const updateCaptureAnswers = async (capture: Capture): Promise<Capture> => {
  return await useAccount(async ({ customer, auth, customerConsents }) => {
    return await useService(async ({ accountService, accessModel, sandbox = true }) => {
      const response = await accountService.updateCaptureAnswers({ customer, ...capture }, sandbox, auth.jwt);

      if (response.errors.length > 0) throw new Error(response.errors[0]);

      await afterLogin(auth, response.responseData as Customer, customerConsents, accessModel, false);

      return response.responseData;
    });
  });
};

export const resetPassword = async (email: string, resetUrl: string) => {
  return await useService(async ({ accountService, sandbox = true, authProviderId }) => {
    const response = await accountService.resetPassword(
      {
        customerEmail: email,
        publisherId: authProviderId,
        resetUrl,
      },
      sandbox,
    );

    if (response.errors.length > 0) throw new Error(response.errors[0]);

    return response.responseData;
  });
};

export const changePasswordWithOldPassword = async (oldPassword: string, newPassword: string, newPasswordConfirmation: string) => {
  return await useService(async ({ accountService, sandbox = true }) => {
    const response = await accountService.changePasswordWithOldPassword({ oldPassword, newPassword, newPasswordConfirmation }, sandbox);
    if (response?.errors?.length > 0) throw new Error(response.errors[0]);

    return response?.responseData;
  });
};

export const changePasswordWithToken = async (customerEmail: string, newPassword: string, resetPasswordToken: string, newPasswordConfirmation: string) => {
  return await useService(async ({ accountService, sandbox = true, authProviderId }) => {
    const response = await accountService.changePasswordWithResetToken(
      { publisherId: authProviderId, customerEmail, newPassword, resetPasswordToken, newPasswordConfirmation },
      sandbox,
    );
    if (response?.errors?.length > 0) throw new Error(response.errors[0]);

    return response?.responseData;
  });
};

export const updateSubscription = async (status: 'active' | 'cancelled'): Promise<unknown> => {
  return await useAccount(async ({ customerId, auth: { jwt } }) => {
    return await useService(async ({ subscriptionService, sandbox = true }) => {
      if (!subscriptionService) throw new Error('subscription service is not configured');
      const { subscription } = useAccountStore.getState();
      if (!subscription) throw new Error('user has no active subscription');

      const response = await subscriptionService?.updateSubscription(
        {
          customerId,
          offerId: subscription.offerId,
          status,
          unsubscribeUrl: subscription.unsubscribeUrl,
        },
        sandbox,
        jwt,
      );

      if (response.errors.length > 0) throw new Error(response.errors[0]);

      await reloadActiveSubscription({ delay: 2000 });

      return response?.responseData;
    });
  });
};

export async function checkEntitlements(offerId?: string): Promise<unknown> {
  return await useAccount(async ({ auth: { jwt } }) => {
    return await useService(async ({ checkoutService, sandbox = true }) => {
      if (!checkoutService) throw new Error('checkout service is not configured');
      if (!offerId) {
        return false;
      }
      const { responseData } = await checkoutService.getEntitlements({ offerId }, sandbox, jwt);
      return !!responseData?.accessGranted;
    });
  });
}

export async function reloadActiveSubscription({ delay }: { delay: number } = { delay: 0 }): Promise<unknown> {
  useAccountStore.setState({ loading: true });
  return await useAccount(async ({ customerId, auth: { jwt } }) => {
    return await useService(async ({ subscriptionService, sandbox = true, config }) => {
      if (!subscriptionService) throw new Error('subscription service is not configured');
      // The subscription data takes a few seconds to load after it's purchased,
      // so here's a delay mechanism to give it time to process
      if (delay > 0) {
        return new Promise((resolve: (value?: unknown) => void) => {
          window.setTimeout(() => {
            reloadActiveSubscription().finally(resolve);
          }, delay);
        });
      }
      const [activeSubscription, transactions, activePayment] = await Promise.all([
        subscriptionService.getActiveSubscription({ sandbox, customerId, jwt, config }),
        subscriptionService.getAllTransactions({ sandbox, customerId, jwt }),
        subscriptionService.getActivePayment({ sandbox, customerId, jwt }),
      ]);
      // this invalidates all entitlements caches which makes the useEntitlement hook to verify the entitlements.
      await queryClient.invalidateQueries('entitlements');

      useAccountStore.setState({
        subscription: activeSubscription,
        loading: false,
        transactions,
        activePayment,
      });
    });
  });
}

/**
 * Get multiple media items for the given IDs. This function uses watchlists to get several medias via just one request.
 *
 * @param watchlistId
 * @param mediaIds
 */
export async function getMediaItems(watchlistId: string | undefined | null, mediaIds: string[]) {
  if (!watchlistId) {
    throw new Error('No watchlist defined');
  }

  return getMediaByWatchlist(watchlistId, mediaIds);
}

async function afterLogin(
  auth: AuthData,
  user: Customer,
  customerConsents: CustomerConsent[] | null,
  accessModel: string,
  shouldSubscriptionReload: boolean = true,
) {
  useAccountStore.setState({
    auth,
    user,
    customerConsents,
  });

  return await Promise.allSettled([
    accessModel === 'SVOD' && shouldSubscriptionReload ? reloadActiveSubscription() : Promise.resolve(),
    getPublisherConsents(),
  ]);
}
