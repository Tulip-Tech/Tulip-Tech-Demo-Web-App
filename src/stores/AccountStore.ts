import type { PaymentDetail, Subscription, Transaction } from '#types/subscription';
import type { AuthData, Consent, Customer, CustomerConsent } from '#types/account';
import { createStore } from '#src/stores/utils';

type AccountStore = {
  loading: boolean;
  auth: AuthData | null;
  user: Customer | null;
  subscription: Subscription | null;
  transactions: Transaction[] | null;
  activePayment: PaymentDetail | null;
  customerConsents: CustomerConsent[] | null;
  publisherConsents: Consent[] | null;
  canUpdateEmail: boolean;
  canRenewSubscription: boolean;
  canChangePasswordWithOldPassword: boolean;
  setLoading: (loading: boolean) => void;
};

export const useAccountStore = createStore<AccountStore>('AccountStore', (set) => ({
  loading: true,
  auth: null,
  user: null,
  subscription: null,
  transactions: null,
  activePayment: null,
  customerConsents: null,
  publisherConsents: null,
  canUpdateEmail: false,
  canRenewSubscription: false,
  canChangePasswordWithOldPassword: false,
  setLoading: (loading: boolean) => set({ loading }),
}));
