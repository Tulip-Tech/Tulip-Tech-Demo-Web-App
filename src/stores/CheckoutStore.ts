import type { Offer, Order, PaymentMethod } from '#types/checkout';
import type { MediaOffer } from '#types/media';
import { createStore } from '#src/stores/utils';

type CheckoutStore = {
  offer: Offer | null;
  order: Order | null;
  paymentMethods: PaymentMethod[] | null;
  requestedMediaOffers: MediaOffer[] | null;
  setOffer: (offer: Offer | null) => void;
  setOrder: (order: Order | null) => void;
  setPaymentMethods: (paymentMethods: PaymentMethod[] | null) => void;
  setRequestedMediaOffers: (requestedMediaOffers: MediaOffer[] | null) => void;
};

export const useCheckoutStore = createStore<CheckoutStore>('CheckoutStore', (set) => ({
  offer: null,
  order: null,
  paymentMethods: null,
  requestedMediaOffers: null,
  setOffer: (offer) => set({ offer }),
  setOrder: (order) => set({ order }),
  setPaymentMethods: (paymentMethods) => set({ paymentMethods }),
  setRequestedMediaOffers: (requestedMediaOffers) => set({ requestedMediaOffers }),
}));
