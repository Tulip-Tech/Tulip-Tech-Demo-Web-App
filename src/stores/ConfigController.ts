import { useConfigStore } from '#src/stores/ConfigStore';
import { getAdSchedule } from '#src/services/api.service';

export const initializeAdSchedule = async () => {
  const { config } = useConfigStore.getState();

  const adScheduleData = await getAdSchedule(config?.adSchedule);

  useConfigStore.setState({
    adScheduleData,
  });
};
