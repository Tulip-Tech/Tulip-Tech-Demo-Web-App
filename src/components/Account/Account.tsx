import React, { useMemo } from 'react';
import { useTranslation } from 'react-i18next';
import { useLocation, useNavigate } from 'react-router-dom';
import shallow from 'zustand/shallow';
import DOMPurify from 'dompurify';

import type { FormSectionContentArgs, FormSectionProps } from '#components/Form/FormSection';
import Visibility from '#src/icons/Visibility';
import VisibilityOff from '#src/icons/VisibilityOff';
import Button from '#components/Button/Button';
import Form from '#components/Form/Form';
import IconButton from '#components/IconButton/IconButton';
import TextField from '#components/TextField/TextField';
import Checkbox from '#components/Checkbox/Checkbox';
import HelperText from '#components/HelperText/HelperText';
import useToggle from '#src/hooks/useToggle';
import { formatConsentsFromValues, formatConsentValues } from '#src/utils/collection';
import { addQueryParam } from '#src/utils/location';
import { useAccountStore } from '#src/stores/AccountStore';
import { logDev } from '#src/utils/common';
import { updateConsents, updateUser } from '#src/stores/AccountController';

type Props = {
  panelClassName?: string;
  panelHeaderClassName?: string;
  canUpdateEmail?: boolean;
};

interface FormErrors {
  email?: string;
  confirmationPassword?: string;
  firstName?: string;
  lastName?: string;
  form?: string;
}

const Account = ({ panelClassName, panelHeaderClassName, canUpdateEmail = true }: Props): JSX.Element => {
  const { t } = useTranslation('user');
  const navigate = useNavigate();
  const location = useLocation();
  const [viewPassword, toggleViewPassword] = useToggle();

  const { customer, customerConsents, publisherConsents, canChangePasswordWithOldPassword } = useAccountStore(
    ({ user, customerConsents, publisherConsents, canChangePasswordWithOldPassword }) => ({
      customer: user,
      customerConsents,
      publisherConsents,
      canChangePasswordWithOldPassword,
    }),
    shallow,
  );

  const consentValues = useMemo(() => formatConsentValues(publisherConsents, customerConsents), [publisherConsents, customerConsents]);

  const initialValues = useMemo(
    () => ({
      ...customer,
      consents: consentValues,
      confirmationPassword: '',
    }),
    [customer, consentValues],
  );

  const formatConsentLabel = (label: string): string | JSX.Element => {
    const sanitizedLabel = DOMPurify.sanitize(label);
    const hasHrefOpenTag = /<a(.|\n)*?>/.test(sanitizedLabel);
    const hasHrefCloseTag = /<\/a(.|\n)*?>/.test(sanitizedLabel);

    if (hasHrefOpenTag && hasHrefCloseTag) {
      return <span dangerouslySetInnerHTML={{ __html: label }} />;
    }

    return label;
  };

  function translateErrors(errors?: string[]): FormErrors {
    const formErrors: FormErrors = {};
    // Some errors are combined in a single CSV string instead of one string per error
    errors
      ?.flatMap((e) => e.split(','))
      .forEach((error) => {
        switch (error.trim()) {
          case 'Invalid param email':
            formErrors.email = t('account.errors.invalid_param_email');
            break;
          case 'Customer email already exists':
            formErrors.email = t('account.errors.email_exists');
            break;
          case 'Please enter a valid e-mail address.':
            formErrors.email = t('account.errors.please_enter_valid_email');
            break;
          case 'Invalid confirmationPassword': {
            formErrors.confirmationPassword = t('account.errors.invalid_password');
            break;
          }
          case 'firstName can have max 50 characters.': {
            formErrors.firstName = t('account.errors.first_name_too_long');
            break;
          }
          case 'lastName can have max 50 characters.': {
            formErrors.lastName = t('account.errors.last_name_too_long');
            break;
          }
          case 'Email update not supported': {
            formErrors.form = t('account.errors.email_update_not_supported');
            break;
          }
          default: {
            formErrors.form = t('account.errors.unknown_error');
            logDev('Unknown error', error);
            break;
          }
        }
      });

    return formErrors;
  }

  function formSection(props: FormSectionProps<typeof initialValues, FormErrors>) {
    return {
      ...props,
      className: panelClassName,
      panelHeaderClassName: panelHeaderClassName,
      saveButton: t('account.save'),
      cancelButton: t('account.cancel'),
      content: (args: FormSectionContentArgs<typeof initialValues, string[]>) => {
        // This function just allows the sections below to use the FormError type instead of an array of errors
        const formErrors = translateErrors(args.errors);

        // Render the section content, but also add a warning text if there's a form level error
        return (
          <>
            {props.content({ ...args, errors: formErrors })}
            <HelperText error={!!formErrors?.form}>{formErrors?.form}</HelperText>
          </>
        );
      },
    };
  }

  const editPasswordClickHandler = () => {
    const modal = canChangePasswordWithOldPassword ? 'edit-password' : 'reset-password';
    navigate(addQueryParam(location, 'u', modal));
  };

  return (
    <Form initialValues={initialValues}>
      {[
        formSection({
          label: t('account.email'),
          onSubmit: (values) =>
            updateUser({
              email: values.email || '',
              confirmationPassword: values.confirmationPassword,
            }),
          canSave: (values) => !!(values.email && values.confirmationPassword),
          editButton: t('account.edit_account'),
          readOnly: !canUpdateEmail,
          content: (section) => (
            <>
              <TextField
                name="email"
                label={t('account.email')}
                value={section.values.email || ''}
                onChange={section.onChange}
                error={!!section.errors?.email}
                helperText={section.errors?.email}
                disabled={section.isBusy}
                editing={section.isEditing}
                required
              />
              {section.isEditing && (
                <TextField
                  name="confirmationPassword"
                  label={t('account.confirm_password')}
                  value={section.values.confirmationPassword}
                  onChange={section.onChange}
                  error={!!section.errors?.confirmationPassword}
                  helperText={section.errors?.confirmationPassword}
                  type={viewPassword ? 'text' : 'password'}
                  disabled={section.isBusy}
                  rightControl={
                    <IconButton aria-label={viewPassword ? t('account.hide_password') : t('account.view_password')} onClick={() => toggleViewPassword()}>
                      {viewPassword ? <Visibility /> : <VisibilityOff />}
                    </IconButton>
                  }
                  required
                />
              )}
            </>
          ),
        }),
        formSection({
          label: t('account.security'),
          editButton: <Button label={t('account.edit_password')} type="button" onClick={() => (customer ? editPasswordClickHandler() : null)} />,
          content: () => (
            <>
              <strong>{t('account.password')}</strong>
              <p>****************</p>
            </>
          ),
        }),
        formSection({
          label: t('account.about_you'),
          editButton: t('account.edit_information'),
          onSubmit: (values) => updateUser({ firstName: values.firstName || '', lastName: values.lastName || '' }),
          content: (section) => (
            <>
              <TextField
                name="firstName"
                label={t('account.firstname')}
                value={section.values.firstName || ''}
                onChange={section.onChange}
                error={!!section.errors?.firstName}
                helperText={section.errors?.firstName}
                disabled={section.isBusy}
                editing={section.isEditing}
              />
              <TextField
                name="lastName"
                label={t('account.lastname')}
                value={section.values.lastName || ''}
                onChange={section.onChange}
                error={!!section.errors?.lastName}
                helperText={section.errors?.lastName}
                disabled={section.isBusy}
                editing={section.isEditing}
              />
            </>
          ),
        }),
        formSection({
          label: t('account.terms_and_tracking'),
          saveButton: t('account.update_consents'),
          onSubmit: (values) => updateConsents(formatConsentsFromValues(publisherConsents, values)),
          content: (section) => (
            <>
              {publisherConsents?.map((consent, index) => (
                <Checkbox
                  key={index}
                  name={`consents.${consent.name}`}
                  value={consent.value || ''}
                  checked={(section.values.consents?.[consent.name] as boolean) || false}
                  onChange={section.onChange}
                  label={formatConsentLabel(consent.label)}
                  disabled={consent.required || section.isBusy}
                />
              ))}
            </>
          ),
        }),
      ]}
    </Form>
  );
};

export default Account;
