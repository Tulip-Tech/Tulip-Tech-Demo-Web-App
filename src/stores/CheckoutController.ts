import { subscribeToNotifications } from './NotificationsController';

import type { CardPaymentData, CreateOrderArgs, Offer, Order, PaymentMethod, PaymentWithPayPalResponse, UpdateOrderPayload } from '#types/checkout';
import { useCheckoutStore } from '#src/stores/CheckoutStore';
import useAccount from '#src/hooks/useAccount';
import useService from '#src/hooks/useService';

export const createOrder = async (offer: Offer, paymentMethodId?: number): Promise<unknown> => {
  return await useAccount(async ({ customer, auth: { jwt } }) => {
    return await useService(async ({ checkoutService, sandbox = true, authProviderId }) => {
      if (!authProviderId) throw new Error('auth provider is not configured');
      if (!checkoutService) throw new Error('checkout service is not available');

      const createOrderArgs: CreateOrderArgs = {
        offer,
        customerId: customer.id,
        country: customer?.country || '',
        customerIP: customer?.lastUserIp || '',
        paymentMethodId,
      };

      const response = await checkoutService.createOrder(createOrderArgs, sandbox, jwt);

      if (response?.errors?.length > 0) {
        useCheckoutStore.getState().setOrder(null);

        throw new Error(response?.errors[0]);
      }

      useCheckoutStore.getState().setOrder(response.responseData?.order);
    });
  });
};

export const updateOrder = async (order: Order, paymentMethodId?: number, couponCode?: string | null): Promise<unknown> => {
  return await useAccount(async ({ auth: { jwt } }) => {
    return await useService(async ({ checkoutService, sandbox = true, authProviderId }) => {
      if (!authProviderId) throw new Error('auth provider is not configured');
      if (!checkoutService) throw new Error('checkout service is not available');

      const updateOrderPayload: UpdateOrderPayload = {
        order,
        paymentMethodId,
        couponCode,
      };

      const response = await checkoutService.updateOrder(updateOrderPayload, sandbox, jwt);
      if (response.errors.length > 0) {
        // clear the order when the order doesn't exist on the server
        if (response.errors[0].includes(`Order with ${order.id} not found`)) {
          useCheckoutStore.getState().setOrder(null);
        }

        throw new Error(response.errors[0]);
      }
      if (response.responseData?.order) {
        useCheckoutStore.getState().setOrder(response.responseData?.order);
      }
    });
  });
};

export const getPaymentMethods = async (): Promise<PaymentMethod[]> => {
  return await useAccount(async ({ auth: { jwt } }) => {
    return await useService(async ({ checkoutService, sandbox = true }) => {
      if (!checkoutService) throw new Error('checkout service is not available');

      const { paymentMethods } = useCheckoutStore.getState();

      if (paymentMethods) return paymentMethods; // already fetched payment methods

      const response = await checkoutService.getPaymentMethods(sandbox, jwt);

      if (response.errors.length > 0) throw new Error(response.errors[0]);

      useCheckoutStore.getState().setPaymentMethods(response.responseData?.paymentMethods);

      return response.responseData?.paymentMethods;
    });
  });
};

export const paymentWithoutDetails = async (): Promise<unknown> => {
  return await useAccount(async ({ auth: { jwt } }) => {
    return await useService(async ({ checkoutService, sandbox = true, authProviderId }) => {
      if (!checkoutService) throw new Error('checkout service is not available');

      const { order } = useCheckoutStore.getState();

      if (!order) throw new Error('No order created');
      if (!authProviderId) throw new Error('auth provider is not configured');

      const response = await checkoutService.paymentWithoutDetails({ orderId: order.id }, sandbox, jwt);

      if (response.errors.length > 0) throw new Error(response.errors[0]);
      if (response.responseData.rejectedReason) throw new Error(response.responseData.rejectedReason);

      return response.responseData;
    });
  });
};

export const directPostCardPayment = async (cardPaymentPayload: CardPaymentData): Promise<unknown> => {
  return await useAccount(async ({ customer }) => {
    return await useService(async ({ checkoutService, authProviderId }) => {
      const { order } = useCheckoutStore.getState();

      if (!order) throw new Error('No order created');
      if (!authProviderId) throw new Error('auth provider is not configured');
      if (!checkoutService) throw new Error('checkout service is not available');

      // subscribe to listen to inplayer websocket notifications
      await subscribeToNotifications(customer?.uuid);

      const response = await checkoutService.directPostCardPayment(cardPaymentPayload, order);

      return response;
    });
  });
};

export const iFrameCardPayment = async (paymentMethod: AdyenPaymentMethod): Promise<unknown> => {
  return await useAccount(async ({ auth: { jwt } }) => {
    return await useService(async ({ checkoutService, sandbox = true, authProviderId }) => {
      const { order } = useCheckoutStore.getState();

      if (!order) throw new Error('No order created');
      if (!authProviderId) throw new Error('auth provider is not configured');
      if (!checkoutService) throw new Error('checkout service is not available');

      const response = await checkoutService.iFrameCardPayment(
        {
          orderId: order.id,
          card: paymentMethod,
        },
        sandbox,
        jwt,
      );

      if (response.errors.length > 0) throw new Error(response.errors[0]);
      if (response.responseData.rejectedReason) throw new Error(response.responseData.rejectedReason);

      return response.responseData;
    });
  });
};

export const paypalPayment = async (successUrl: string, cancelUrl: string, errorUrl: string, couponCode: string = ''): Promise<PaymentWithPayPalResponse> => {
  return await useAccount(async ({ auth: { jwt } }) => {
    return await useService(async ({ checkoutService, sandbox = true, authProviderId }) => {
      const { order } = useCheckoutStore.getState();

      if (!order) throw new Error('No order created');
      if (!authProviderId) throw new Error('auth provider is not configured');
      if (!checkoutService) throw new Error('checkout service is not available');

      const response = await checkoutService.paymentWithPayPal(
        {
          order: order,
          successUrl,
          cancelUrl,
          errorUrl,
          couponCode,
        },
        sandbox,
        jwt,
      );

      if (response.errors.length > 0) throw new Error(response.errors[0]);

      return response.responseData;
    });
  });
};
