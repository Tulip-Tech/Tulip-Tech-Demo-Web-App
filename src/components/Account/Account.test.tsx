import React from 'react';

import Account from './Account';

import customer from '#test/fixtures/customer.json';
import { useAccountStore } from '#src/stores/AccountStore';
import { renderWithRouter } from '#test/testUtils';
import type { Consent } from '#types/account';

describe('<Account>', () => {
  test('renders and matches snapshot', () => {
    useAccountStore.setState({
      user: customer,
      publisherConsents: Array.of({ name: 'marketing', label: 'Receive Marketing Emails' } as Consent),
    });

    const { container } = renderWithRouter(<Account panelClassName={'panel-class'} panelHeaderClassName={'header-class'} canUpdateEmail={true} />);

    // todo
    expect(container).toMatchSnapshot();
  });
});
