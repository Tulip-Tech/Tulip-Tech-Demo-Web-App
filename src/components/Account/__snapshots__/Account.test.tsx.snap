// Vitest Snapshot v1

exports[`<Account> > renders and matches snapshot 1`] = `
<div>
  <div
    class="panel-class"
  >
    <div
      class="header-class"
    >
      <h3>
        account.email
      </h3>
    </div>
    <div
      class="_flexBox_1c1c63"
    >
      <div
        class="_textField_e16c1b"
      >
        <label
          class="_label_e16c1b"
          for="text-field_1235_email"
        >
          account.email
        </label>
        <p>
          email@domain.com
        </p>
      </div>
    </div>
    <div
      class="_controls_1c1c63"
    >
      <button
        class="_button_f8f296 _default_f8f296 _outlined_f8f296"
        type="button"
      >
        <span>
          account.edit_account
        </span>
      </button>
    </div>
  </div>
  <div
    class="panel-class"
  >
    <div
      class="header-class"
    >
      <h3>
        account.security
      </h3>
    </div>
    <div
      class="_flexBox_1c1c63"
    >
      <strong>
        account.password
      </strong>
      <p>
        ****************
      </p>
    </div>
    <div
      class="_controls_1c1c63"
    >
      <button
        class="_button_f8f296 _default_f8f296 _outlined_f8f296"
        type="button"
      >
        <span>
          account.edit_password
        </span>
      </button>
    </div>
  </div>
  <div
    class="panel-class"
  >
    <div
      class="header-class"
    >
      <h3>
        account.about_you
      </h3>
    </div>
    <div
      class="_flexBox_1c1c63"
    >
      <div
        class="_textField_e16c1b"
      >
        <label
          class="_label_e16c1b"
          for="text-field_1235_firstName"
        >
          account.firstname
        </label>
        <p>
          John
        </p>
      </div>
      <div
        class="_textField_e16c1b"
      >
        <label
          class="_label_e16c1b"
          for="text-field_1235_lastName"
        >
          account.lastname
        </label>
        <p>
          Doe
        </p>
      </div>
    </div>
    <div
      class="_controls_1c1c63"
    >
      <button
        class="_button_f8f296 _default_f8f296 _outlined_f8f296"
        type="button"
      >
        <span>
          account.edit_information
        </span>
      </button>
    </div>
  </div>
  <div
    class="panel-class"
  >
    <div
      class="header-class"
    >
      <h3>
        account.terms_and_tracking
      </h3>
    </div>
    <div
      class="_flexBox_1c1c63"
    >
      <div
        class="_checkbox_531f07"
      >
        <div
          class="_row_531f07"
        >
          <input
            id="check-box_1235_consents.marketing"
            name="consents.marketing"
            type="checkbox"
            value=""
          />
          <label
            for="check-box_1235_consents.marketing"
          >
            
            Receive Marketing Emails
          </label>
        </div>
      </div>
    </div>
    <div
      class="_controls_1c1c63"
    />
  </div>
</div>
`;
